Automated Smoketest: :x:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 834451a68326352fbc01bb289271ddbce1fa6696
Number of Error(s): 111
/mnt/disk1/home/vamosax/smoketest/PR-11965/HPCC-Platform/configuration/configmgr/configmgrlib/SchemaValue.hpp:95:1: error: expected unqualified-id before \xe2\x80\x98<<\xe2\x80\x99 token

[Build error:](https://gist.github.com/832cc49efae974a955953653569a6814)
Build log [Build log](https://gist.github.com/HPCCSmoketest/832cc49efae974a955953653569a6814#file-result-18-12-10-15-15-27-log)

--- a/configuration/configmgr/configmgrlib/SchemaItem.cpp
+++ b/configuration/configmgr/configmgrlib/SchemaItem.cpp
@@ -66,10 +66,9 @@ SchemaItem::SchemaItem(const SchemaItem &item)
 
     //
     // Make a copy of the children now
-    for (auto childIt = item.m_children.begin(); childIt != item.m_children.end(); ++childIt)
+    for (auto &pChild: m_children)
     {
-        std::shared_ptr<SchemaItem> pChild = std::make_shared<SchemaItem>((*(childIt->second)));
-        addChild(pChild);
+        addChild(std::make_shared<SchemaItem>(*pChild));
     }
 
     //
@@ -235,13 +234,14 @@ void SchemaItem::setParentForChildren(std::shared_ptr<SchemaItem> pParent)
     setParent(pParent);
     for (auto &childIt: m_children)
     {
-        childIt.second->setParentForChildren(shared_from_this());
+        childIt->setParentForChildren(shared_from_this());
     }
 }
 
 
 void SchemaItem::addAttribute(const std::shared_ptr<SchemaValue> &pCfgValue)
 {
+    pCfgValue->setOrdinal(m_attributes.size());
     auto retVal = m_attributes.insert({ pCfgValue->getName(), pCfgValue });
     if (!retVal.second)
     {
@@ -352,13 +352,13 @@ void SchemaItem::processUniqueAttributeValueSetReferences(const std::map<std::st
 
 void SchemaItem::getChildren(std::vector<std::shared_ptr<SchemaItem>> &children, const std::string &name, const std::string &itemType) const
 {
-    for (auto it = m_children.begin(); it != m_children.end(); ++it)
+    for (auto &pChild: m_children)
     {
-        if (name.empty() || it->first == name)
+        if (name.empty() || pChild->getProperty("name") == name)
         {
-            if (itemType.empty() || it->second->getProperty("itemType") == itemType)
+            if (itemType.empty() || pChild->getProperty("itemType") == itemType)
             {
-                children.emplace_back(it->second);
+                children.emplace_back(pChild);
             }
         }
     }
@@ -535,9 +535,9 @@ void SchemaItem::processDefinedUniqueAttributeValueSets(std::map<std::string, st
 
     //
     // Post process all of our children now
-    for (auto it = m_children.begin(); it != m_children.end(); ++it)
+    for (auto &pChild: m_children)
     {
-        it->second->processDefinedUniqueAttributeValueSets(uniqueAttributeValueSets);
+        pChild->processDefinedUniqueAttributeValueSets(uniqueAttributeValueSets);
     }
 }
 
@@ -547,14 +547,14 @@ void SchemaItem::postProcessConfig(const std::map<std::string, std::vector<std::
     //
     // Make sure that the item type value for all children that are insertable (minRequired = 0 or maxAllowed > minRequired)
     std::set<std::string> itemTypes;
-    for (auto it = m_children.begin(); it != m_children.end(); ++it)
+    for (auto &pChild: m_children)
     {
-        if (it->second->isInsertable())
+        if (pChild->isInsertable())
         {
-            auto rc = itemTypes.insert(it->second->getItemType());
+            auto rc = itemTypes.insert(pChild->getItemType());
             if (!rc.second)
             {
-                throw(ParseException("Element: " + getProperty("name") + ", duplicate itemType(" + it->second->getItemType() + ") found"));
+                throw(ParseException("Element: " + getProperty("name") + ", duplicate itemType(" + pChild->getItemType() + ") found"));
             }
         }
     }
@@ -592,9 +592,9 @@ void SchemaItem::postProcessConfig(const std::map<std::string, std::vector<std::
 
     //
     // Post process all of our children now
-    for (auto it = m_children.begin(); it!= m_children.end(); ++it)
+    for (auto &pChild: m_children)
     {
-        it->second->postProcessConfig(uniqueAttributeValueSets);
+        pChild->postProcessConfig(uniqueAttributeValueSets);
     }
 }
 

--- a/configuration/configmgr/configmgrlib/SchemaItem.hpp
+++ b/configuration/configmgr/configmgrlib/SchemaItem.hpp
@@ -50,8 +50,7 @@ class DECL_EXPORT SchemaItem : public std::enable_shared_from_this<SchemaItem>
         void addSchemaType(const std::shared_ptr<SchemaItem> &pItem, const std::string &typeName);
         std::shared_ptr<SchemaItem> getSchemaType(const std::string &name, bool throwIfNotPresent=true) const;
         void insertSchemaType(const std::shared_ptr<SchemaItem> pTypeItem);
-        void addChild(const std::shared_ptr<SchemaItem> &pItem) { m_children.insert({ pItem->getProperty("name"), pItem }); }
-        void addChild(const std::shared_ptr<SchemaItem> &pItem, const std::string &name) { m_children.insert({ name, pItem }); }
+        void addChild(const std::shared_ptr<SchemaItem> &pItem) { m_children.emplace_back(pItem); }
         void getChildren(std::vector<std::shared_ptr<SchemaItem>> &children, const std::string &name = std::string(""), const std::string &itemType = std::string("")) const;
         void setItemSchemaValue(const std::shared_ptr<SchemaValue> &pValue) { m_pItemValue = pValue; }
         std::shared_ptr<SchemaValue> getItemSchemaValue() const { return m_pItemValue; }
@@ -99,7 +98,7 @@ class DECL_EXPORT SchemaItem : public std::enable_shared_from_this<SchemaItem>
         bool m_hidden;
         unsigned m_minInstances;
         unsigned m_maxInstances;
-        std::multimap<std::string, std::shared_ptr<SchemaItem>> m_children;
+        std::vector<std::shared_ptr<SchemaItem>> m_children;
         std::shared_ptr<SchemaValue> m_pItemValue;   // value for this item (think of it as the VALUE for an element <xx attr= att1>VALUE</xx>)
         std::map<std::string, std::shared_ptr<SchemaValue>> m_attributes;   // attributes for this item (think in xml terms <m_name attr1="val" attr2="val" .../> where attrN is in this vector
         std::set<std::string> m_keys;   // generic set of key values for use by any component to prevent duplicate operations
@@ -127,7 +126,6 @@ class DECL_EXPORT SchemaItem : public std::enable_shared_from_this<SchemaItem>
 
         // Following are NOT copied in copy constructor
         std::weak_ptr<SchemaItem> m_pParent;
-
 };
 
 #endif // _CONFIG2_CONFIGITEM_HPP_

--- a/configuration/configmgr/configmgrlib/SchemaValue.cpp
+++ b/configuration/configmgr/configmgrlib/SchemaValue.cpp
@@ -33,6 +33,7 @@ SchemaValue::SchemaValue(const std::string &name, bool isDefined) :
     bitMask.m_isUnique = 0;
     bitMask.m_isDefined = isDefined;
     bitMask.m_noOutput = 0;
+    m_ordinal = 0;
 }
 
 
@@ -60,6 +61,8 @@ SchemaValue::SchemaValue(const SchemaValue &value)
     // special processing? Maybe after inserting?
     std::vector<std::shared_ptr<SchemaValue>> m_mirrorToSchemaValues;
     std::vector<std::weak_ptr<SchemaValue>> m_pUniqueValueSetRefs;    // this value serves as the key from which values are valid
+
+    m_ordinal = 0;  // assigned when inserted
 }
 
 

--- a/configuration/configmgr/configmgrlib/SchemaValue.hpp
+++ b/configuration/configmgr/configmgrlib/SchemaValue.hpp
@@ -92,6 +92,8 @@ class DECL_EXPORT SchemaValue
         const std::string &getRequiredIf() const { return m_requiredIf; }
         void setGroupByName(const std::string &group) { m_groupByName = group; }
         const std::string &getGroupByName() const { return m_groupByName; }
+        void setOrdinal(unsigned val) { m_ordinal = val; }
+        unsigned getOrdinal() const { return m_ordinal; }
         void setNoOutput(bool noOutput) { bitMask.m_noOutput = noOutput; }
         bool isNoOutput() const { return bitMask.m_noOutput; }
 
@@ -104,6 +106,7 @@ class DECL_EXPORT SchemaValue
     protected:
 
         // DON'T FORGET IF DATA ADDED, IT MAY MAY TO BE COPIED IN THE COPY CONSTRUCTOR!!
+        unsigned m_ordinal;
         std::shared_ptr<SchemaType> m_pType;
         std::string m_name;
         std::string m_displayName;

--- a/esp/scm/ws_configmgr.ecm
+++ b/esp/scm/ws_configmgr.ecm
@@ -202,6 +202,7 @@ ESPstruct AttributeType
 {
     string     DisplayName;
     string     Name;
+    unsigned   Ordinal;
     string     Group("");
     string     Tooltip("");
     ESPstruct  TypeInfo Type;

--- a/esp/services/ws_configmgr/ws_configmgrService.cpp
+++ b/esp/services/ws_configmgr/ws_configmgrService.cpp
@@ -868,6 +868,11 @@ void Cws_configMgrEx::getAttributes(const std::shared_ptr<EnvironmentNode> &pEnv
     std::vector<std::shared_ptr<SchemaValue>> schemaValues;
     pEnvNode->getSchemaItem()->getAttributes(schemaValues);
 
+    //
+    // Because the config manager creates a schema value for ALL attributes of a node (even those that are not explicitly defined
+    // in the schema), all attributes of the environment node are processed. Additionally, if the environment node is missing
+    // attributes that are defined in the schema, they are found and returned as missing (if indicated by the includeMissing
+    // parameter)
     for (auto it=schemaValues.begin(); it!=schemaValues.end(); ++it)
     {
         const std::shared_ptr<SchemaValue> pSchemaValue = *it;
@@ -879,6 +884,7 @@ void Cws_configMgrEx::getAttributes(const std::shared_ptr<EnvironmentNode> &pEnv
             pAttribute->setName(pSchemaValue->getName().c_str());
             pAttribute->setDisplayName(pSchemaValue->getDisplayName().c_str());
             pAttribute->setTooltip(pSchemaValue->getTooltip().c_str());
+            pAttribute->setOrdinal(pSchemaValue->getOrdinal());
 
             const std::shared_ptr<SchemaType> &pType = pSchemaValue->getType();
             std::shared_ptr<SchemaTypeLimits> &pLimits = pType->getLimits();

